src/macros/@production.jl

```diff
@@ -0,0 +1,385 @@
+            elas[i=I,j=J], 4
+            T[i=I], .1
+        end)
+
+julia> @sectors(M, begin
+            X
+            Y[i=I, j=J]
+            Z[i=I]
+        end);
+
+julia> @commodities(M, begin
+            PX
+            PY
+            PL[i=I]
+            PK
+        end);
+
+julia> @consumer(M, RA);
+
+julia> @production(M, X, [s=1,t=0], begin
+            @output(PX, 10, t)
+            @input(PL, 5, s)
+            @input(PK, 5, s)
+        end);
+
+julia> @production(M, Y[i=I,j=J], [s=2, t=1, va[ii=I]=>s=elas[i,ii]], begin
+            @output(PY, 15, t)
+            @input(PX, 3, s)
+            @input(PL[ii=I], 4, va[ii], taxes = [Tax(RA, .5)])
+            @input(PK, 6, va[i])
+        end);
+
+julia> for i in I
+            @production(M, Z[i], [s=1, t=0], begin
+                @output(PK, 20, t)
+                @input(PL[i], 2, s, taxes = [Tax(RA, T[i])])
+            end)
+        end
+```
+"""
+macro production(input_args...)
+    error_fn = Containers.build_error_fn("production", input_args, __source__)
+    args,kwargs = Containers.parse_macro_arguments(
+        error_fn, 
+        input_args; 
+        num_positional_args = 4,
+        valid_kwargs = nothing
+        )
+
+    model = esc(args[1])
+
+    sector, index_vars, indices, all_indices = parse_ref_sets(error_fn, args[2])
+    sector_name_expr = build_name_expr(sector, all_indices, kwargs)
+    
+    nests = build_nest_expr(args[3], __source__, index_vars)
+    netputs = build_netputs(args[4], index_vars)
+
+    build_production_sectors = JuMP.Containers.container_code(
+        index_vars,
+        indices,
+        quote
+            PreProduction(
+                $sector_name_expr,
+                $nests,
+                $netputs
+            )
+        end,
+        :DenseAxisArray
+    )
+   
+    production_code = quote
+        P = build_production(
+            $error_fn,
+            $model,
+            $index_vars,
+            $(esc(sector)),
+            $build_production_sectors
+        )
+       add_production!($model, P)
+    end
+   
+    return production_code
+end
+
+
+function build_production(
+        error_fn::Function,
+        model,
+        index_vars::Vector{Any},
+        base_sector,
+        pre_preduction::PreProduction
+        )
+
+    sector = pre_preduction.sector
+    nests = pre_preduction.nests
+    netputs = pre_preduction.netputs
+
+    return ScalarProduction(
+        error_fn, 
+        sector, 
+        create_nodes(model, nests), 
+        netputs,
+        index_vars
+    )
+end
+
+# Currently returns a dense axis array of scalar productions
+function build_production(
+        error_fn::Function,
+        model,
+        index_vars::Vector{Any},
+        base_sector,
+        pre_production::AbstractArray{<:PreProduction}
+        )
+
+    P =  build_production.(error_fn, Ref(model), Ref(index_vars), Ref(base_sector), pre_production)
+    return IndexedProduction(base_sector, P, index_vars)
+end
```

Comment:
Broadcast over the `pre_production` entry. `Ref` just means "don't broadcast this bit"